File: ga/gpt3_5_result_2024_6/result/1968D_gpt/26_1968D.c
use std::cmp;

const N: usize = 200000;

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();

    for _ in 0..t {
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let k: usize = iter.next().unwrap().parse().unwrap();
        let a: usize = iter.next().unwrap().parse().unwrap();
        let b: usize = iter.next().unwrap().parse().unwrap();
        let m = cmp::min(n, k);
        let (mut a_, mut b_) = (a, b);

        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let arr: Vec<usize> = input.split_whitespace().map(|x| x.parse().unwrap()).collect();

        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let aa: Vec<usize> = input.split_whitespace().map(|x| x.parse().unwrap()).collect();

        let (mut p1, mut p2, mut p1_, mut p2_) = (0, 0, 0, 0);
        let (mut i, mut i_, mut j, mut j_) = (-1, -1, -1, -1);

        for x in 1..=m {
            p1 = cmp::max(p1, aa[a] * (k - x + 1) as i64 + p1_);
            p1_ += aa[a];
            a = arr[a];
            p2 = cmp::max(p2, aa[b] * (k - x + 1) as i64 + p2_);
            p2_ += aa[b];
            b = arr[b];
        }

        println!("{}", if p1 > p2 { "Bodya" } else if p1 < p2 { "Sasha" } else { "Draw" });
    }
}

File: ga/4omini_result_2024_7/result/1968D_mini/26_1968D.rs
use std::cmp::{min, max};
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    
    let mut lines = input.lines();
    let t: usize = lines.next().unwrap().parse().unwrap();
    
    for _ in 0..t {
        let first_line: Vec<usize> = lines.next().unwrap().split_whitespace()
                                             .map(|x| x.parse().unwrap())
                                             .collect();
                                             
        let (n, k, a, b) = (first_line[0], first_line[1], first_line[2], first_line[3]);
        let m = min(n, k);
        
        let arr: Vec<usize> = lines.next().unwrap().split_whitespace()
                                     .map(|x| x.parse().unwrap())
                                     .collect();
                                     
        let aa: Vec<usize> = lines.next().unwrap().split_whitespace()
                                     .map(|x| x.parse().unwrap())
                                     .collect();
                                     
        let mut p1 = 0u64;
        let mut p2 = 0u64;
        let mut p1_ = 0u64;
        let mut p2_ = 0u64;
        
        let mut a_ind = a;
        let mut b_ind = b;
        
        for x in 0..m {
            p1 = max(p1, aa[a_ind] as u64 * (k - x + 1) + p1_);
            p1_ += aa[a_ind] as u64;
            a_ind = arr[a_ind];
            
            p2 = max(p2, aa[b_ind] as u64 * (k - x + 1) + p2_);
            p2_ += aa[b_ind] as u64;
            b_ind = arr[b_ind];
        }
        
        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}